### PR TITLE
To ISIS formatter

### PR DESCRIPTION
**Pull request by jessemapel**
This is almost functional. It is only having issues with converting numpy to JSON right now.

I plan to work on this more tomorrow/Friday, but I want to get it up in case I don't have time or someone wants to work more on it.

### CODE REVIEW

**Comment by jlaura** [comment]
For numpy, I can pull the JSON serializer from plio that handles arrays. I'll take a look at this tonight in case you want to cook on it more between now and then. I don't want to step on your toes with this.

File: ale/formatters/__init__.py [hunk 1]
```diff
@@ -1 +1 @@
-from . import usgscsm_formatter
+from . import usgscsm_formatter, isis_formatter
```
Comment (jessemapel) [inline ale/formatters/__init__.py]: Is this necessary? I'm not familiar enough with `__init__.py` files to know.

File: ale/formatters/isis_formatter.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    meta_data['NaifKeywords'] = driver.isis_naif_keywords
+
+    j2000 = driver.frame_chain
+
+    instrument_pointing = {}
+    sensor_frame = j2000.find_child_frame(driver.sensor_frame_id)
+    time_dependent_sensor_frame = j2000.last_time_dependent_frame_between(sensor_frame)
+    if time_dependent_sensor_frame != j2000:
+        forward_path, reverse_path = j2000.path_to(time_dependent_sensor_frame)
+        instrument_pointing['TimeDependentFrames'] = [frame.id for frame in (forward_path + reverse_path)[::-1]]
```
Comment (jessemapel) [inline ale/formatters/isis_formatter.py]: The reversal here is to conform to ISIS. It has a slightly strange ordering for these.

File: ale/transformation.py [hunk 3]
```diff
@@ -137,3 +164,31 @@ def rotation_to(self, other):
...
+
+        Returns
+        -------
+        FrameNode
+            The first frame between the this frame and the other frame such
+            that the rotation from this frame to the in-between frame is time
+            dependent and the rotation from the in-between frame to the other
+            frame is constant. If there are no time dependent frames between
+            this frame and the other frame, None is returned.
+        """
```
Comment (jessemapel) [inline ale/transformation.py]: I don't know a clean way to name or describe this method. It's a rather complex thing.

File: tests/pytests/test_isis_formatter.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+    def pixel2focal_y(self):
+        return [28, 93, 5]
+
+    @property
+    def ephemeris_start_time(self):
+        return 120
+
+    @property
+    def ephemeris_stop_time(self):
+        return 32
```
Comment (jessemapel) [inline tests/pytests/test_isis_formatter.py]: Should we have 1 set of test drivers that all of the formatters use?

File: tests/pytests/test_transformation.py [hunk 5]
```diff
@@ -125,3 +125,62 @@ def test_self_rotation():
...
+        TimeDependentRotation(
+            np.array([1.0/np.sqrt(2), 0, 0, 1.0/np.sqrt(2)]),
+            np.array([1]), 5, 4),
+        ConstantRotation(np.array([1.0/np.sqrt(2), 0, 0, 1.0/np.sqrt(2)]), 3, 2)
+    ]
+    root_node = FrameNode(1)
+    child_node_1 = FrameNode(2, parent = root_node, rotation = rotations[0])
+    child_node_2 = FrameNode(3, parent = child_node_1, rotation = rotations[1])
+    child_node_3 = FrameNode(4, parent = root_node, rotation = rotations[2])
+    child_node_4 = FrameNode(5, parent = child_node_3, rotation = rotations[3])
```
Comment (jessemapel) [inline tests/pytests/test_transformation.py]: I could see a case for making this the test frame tree used by all of the tests, it makes checking rotations a little more complex though.

Comment (jessemapel) [reply tests/pytests/test_transformation.py]: Also, the node names are confusing, `child_1` is frame `2`, etc.

File: ale/formatters/__init__.py (hunk 1 above)
Comment (jlaura) [reply ale/formatters/__init__.py]: In order to get these into the formatters namespace.

File: ale/formatters/isis_formatter.py (hunk 2 above)
Comment (jlaura) [reply ale/formatters/isis_formatter.py]: Can this comment be inlined into the source.

File: ale/rotation.py [hunk 6]
```diff
@@ -56,6 +56,12 @@ def quat(self, new_quat):
         """
         self._rot = Rotation.from_quat(new_quat)
 
+    def rotation_matrix(self):
+        """
+        The rotation matrix representation of the constant rotation
+        """
+        return self._rot.as_dcm()
+
```
[the remark this answers is not in the document]
Comment (jlaura) [reply ale/rotation.py]: If it does, the test would likely be to simply check that (1) this is being called and (2) the return type is as expected. Any change the return is None? If not, I would argue no. (Put a `  #pragma no cover` into the def line for future code coverage ignores?

File: ale/rotation.py [hunk 7]
```diff
@@ -77,7 +83,7 @@ def __mul__(self, other):
                 Another rotation object, it can be constant or time dependent.
         """
         if self.source != other.dest:
-            raise ValueError("Destination frame of first rotation is not the same as source frame of second rotation.")
+            raise ValueError("Destination frame of first rotation {} is not the same as source frame of second rotation {}.".format(other.dest, self.source))
```
Comment (jlaura) [inline ale/rotation.py]: I have really been liking the new f'{other.dest} blah, blah {self.source}' notation.

File: ale/rotation.py [hunk 8]
```diff
@@ -77,7 +83,7 @@ def __mul__(self, other):
...
         if isinstance(other, ConstantRotation):
             new_rot = self._rot * other._rot
```
Comment (jlaura) [inline ale/rotation.py]: Why are these being treated as private?

File: ale/transformation.py (hunk 3 above)
Comment (jlaura) [reply ale/transformation.py]: I have not reviewed this or down - have to step away for a bit

File: ale/formatters/isis_formatter.py [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+from ale.rotation import ConstantRotation, TimeDependentRotation
+
+def to_isis(driver):
+    meta_data = {}
+
+    meta_data['CameraVersion'] = driver.sensor_model_version
+
+    meta_data['NaifKeywords'] = driver.isis_naif_keywords
+
+    j2000 = driver.frame_chain
```
Comment (SgStapleton) [inline ale/formatters/isis_formatter.py]: I was under the impression that this was being implemented in NaifSpice (https://github.com/USGS-Astrogeology/ale/issues/164 and https://github.com/USGS-Astrogeology/ale/pull/176)?

File: ale/rotation.py (hunk 8 above)
Comment (jessemapel) [reply ale/rotation.py]: `_rot` is the actual implementation of the rotation as a scipy Rotation object. This provides us better encapsulation.

File: ale/formatters/isis_formatter.py (hunk 9 above)
Comment (jessemapel) [reply ale/formatters/isis_formatter.py]: Now that I am looking at an actual implementation, I don't think we need the sensor orientation or anything about rotations besides the frame chain in any of the data classes. The frame chain should has all of the time and rotation information baked into it already.

**Comment by jessemapel** [comment]
Fixes #138

**Comment by jessemapel** [comment]
This is now done and ready to go!

File: tests/pytests/test_isis_formatter.py (hunk 4 above)
Comment (jessemapel) [reply tests/pytests/test_isis_formatter.py]: This can be addressed under #178

**Pull request merged**